In [1]:
import pandas as pd

In [2]:
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")

In [3]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [5]:
customers.shape

(99441, 5)

In [6]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [7]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [8]:
customers.duplicated().sum()

np.int64(0)

In [9]:
customers["customer_id"].nunique()

99441

In [10]:
customers.shape[0]

99441

In [11]:
customers["customer_unique_id"].nunique()

96096

## Customers Dataset — Initial Observations

- The dataset contains 99,441 rows and 5 columns.
- Each row represents a customer record associated with an order.
- `customer_id` is unique for every row and can be used to connect customers with the orders dataset.
- `customer_unique_id` represents the actual customer and may appear multiple times when the same customer places multiple orders.
- The dataset contains no missing values.
- No exact duplicate rows were found.
- Customer location information is available through city, state, and ZIP code prefix.
- For customer-level metrics such as total unique customers, repeat purchase rate, and customer profitability, `customer_unique_id` should be used instead of `customer_id`.

In [14]:
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")

In [15]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [17]:
orders.shape

(99441, 8)

In [18]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [19]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [20]:
orders.duplicated().sum()

np.int64(0)

In [21]:
orders["order_id"].nunique()

99441

In [22]:
orders.shape[0]

99441

In [23]:
orders["customer_id"].nunique()

99441

In [24]:
customers["customer_id"].nunique()

99441

In [25]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [26]:
orders["order_purchase_timestamp"] = pd.to_datetime(
    orders["order_purchase_timestamp"]
)

In [27]:
orders["order_purchase_timestamp"].min()

Timestamp('2016-09-04 21:15:19')

In [28]:
orders["order_purchase_timestamp"].max()

Timestamp('2018-10-17 17:30:18')

In [29]:
orders[~orders["customer_id"].isin(customers["customer_id"])].shape[0]

0

## Orders Dataset — Initial Observations

- The dataset contains 99,441 order records and 8 columns.
- Each row represents one order, making `order_id` the primary key candidate for the Orders table.
- `customer_id` connects the Orders dataset to the Customers dataset.
- Every `customer_id` in Orders has a matching record in the Customers dataset, indicating valid referential integrity between the two tables.
- No exact duplicate rows were found.
- Most orders have a `delivered` status, while a smaller number are canceled, unavailable, shipped, invoiced, processing, created, or approved.
- Missing values exist mainly in approval and delivery timestamp columns.
- These missing values should not be removed automatically because some are associated with legitimate business states such as canceled, unavailable, or undelivered orders.
- The dataset covers purchases from September 2016 to October 2018.
- Before profitability analysis, timestamp columns will need to be converted to datetime data types.
- A business rule will be required to determine which order statuses should be included in revenue and profitability calculations.

In [30]:
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")

In [31]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [32]:
order_items.shape

(112650, 7)

In [33]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [34]:
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [35]:
order_items.duplicated().sum()

np.int64(0)

In [36]:
order_items["order_id"].nunique()

98666

In [37]:
order_items.shape[0]

112650

In [38]:
order_items["order_item_id"].value_counts().sort_index()

order_item_id
1     98666
2      9803
3      2287
4       965
5       460
6       256
7        58
8        36
9        28
10       25
11       17
12       13
13        8
14        7
15        5
16        3
17        3
18        3
19        3
20        3
21        1
Name: count, dtype: int64

In [39]:
order_items.duplicated(
    subset=["order_id", "order_item_id"]
).sum()

np.int64(0)

In [40]:
order_items[["price", "freight_value"]].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


In [41]:
(order_items["price"] <= 0).sum()

np.int64(0)

In [42]:
(order_items["freight_value"] < 0).sum()

np.int64(0)

In [43]:
orders[
    ~orders["order_id"].isin(order_items["order_id"])
].shape[0]

775

In [44]:
order_items.groupby("order_id").size().describe()

count    98666.000000
mean         1.141731
std          0.538452
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         21.000000
dtype: float64

In [45]:
order_items.groupby("order_id").size().sort_values(
    ascending=False
).head(10)

order_id
8272b63d03f5f79c56e9e4120aec44ef    21
1b15974a0141d54e36626dca3fdc731a    20
ab14fdcfbe524636d65ee38360e22ce8    20
9ef13efd6949e4573a18964dd1bbe7f5    15
428a2f660dc84138d969ccd69a0ab6d5    15
9bdc4d4c71aa1de4606060929dee888c    14
73c8ab38f07dc94389065f7eba4f297a    14
37ee401157a3a0b28c9c6d0ed8c3b24b    13
af822dacd6f5cff7376413c03a388bb7    12
3a213fcdfe7d98be74ea0dc05a8b31ae    12
dtype: int64

## Order Items Dataset — Initial Observations

- The dataset contains 112,650 rows and 7 columns.
- Each row represents one item position within an order.
- `order_id` is not unique because one order can contain multiple item rows.
- `order_item_id` is a sequence number within each order and is not globally unique.
- The combination of `order_id` and `order_item_id` uniquely identifies each row, making it a composite key candidate.
- No exact duplicate rows or missing values were found.
- Every `order_id` in the Order Items dataset has a matching record in the Orders dataset, indicating valid referential integrity from Order Items to Orders.
- Some orders do not have corresponding item records. These orders should be investigated by order status before deciding how they should be handled.
- Product prices and freight values are non-negative, although the distributions contain high-value observations that should be investigated before treating them as outliers.
- Most orders contain one item row, while a smaller number contain multiple item rows.
- The Orders to Order Items relationship is one-to-many.
- `price` represents the item selling price, while `freight_value` represents the freight amount associated with that item row.

In [46]:
products = pd.read_csv("../data/raw/olist_products_dataset.csv")

products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [47]:
products.shape

(32951, 9)

In [48]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [49]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [50]:
products.duplicated().sum()

np.int64(0)

In [51]:
products["product_id"].nunique()

32951

In [52]:
products.shape[0]

32951

In [53]:
products["product_id"].nunique()

32951

In [54]:
products["product_category_name"].nunique()

73

In [55]:
products["product_category_name"].value_counts(dropna=False).head(15)

product_category_name
cama_mesa_banho                3029
esporte_lazer                  2867
moveis_decoracao               2657
beleza_saude                   2444
utilidades_domesticas          2335
automotivo                     1900
informatica_acessorios         1639
brinquedos                     1411
relogios_presentes             1329
telefonia                      1134
bebes                           919
perfumaria                      868
papelaria                       849
fashion_bolsas_e_acessorios     849
cool_stuff                      789
Name: count, dtype: int64

In [56]:
products[
    [
        "product_name_lenght",
        "product_description_lenght",
        "product_photos_qty",
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
].describe()

,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32341.000000,32341.000000,32341.000000,32949.000000,32949.000000,32949.000000,32949.000000
mean,48.476949,771.495285,2.188986,2276.472488,30.815078,16.937661,23.196728
std,10.245741,635.115225,1.736766,4282.038731,16.914458,13.637554,12.079047
min,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,57.000000,972.000000,3.000000,1900.000000,38.000000,21.000000,30.000000
max,76.000000,3992.000000,20.000000,40425.000000,105.000000,105.000000,118.000000


In [57]:
order_items[
    ~order_items["product_id"].isin(products["product_id"])
].shape[0]

0

In [58]:
products[
    ~products["product_id"].isin(order_items["product_id"])
].shape[0]

0

In [59]:
order_items.groupby("product_id").size().describe()

count    32951.000000
mean         3.418713
std         10.619709
min          1.000000
25%          1.000000
50%          1.000000
75%          3.000000
max        527.000000
dtype: float64

In [60]:
order_items.groupby("product_id").size().sort_values(
    ascending=False
).head(10)

product_id
aca2eb7d00ea1a7b8ebd4e68314663af    527
99a4788cb24856965c36a24e339b6058    488
422879e10f46682990de24d770e7f83d    484
389d119b48cf3043d311335e499d9c6b    392
368c6c730842d78016ad823897a372db    388
53759a2ecddad2bb87a079a1f1519f73    373
d1c427060a0f73f6b889a5c7c61f2ac4    343
53b36df67ebb7c41585e8d54d6772e08    323
154e7e31ebfa092203795c972e5804a6    281
3dd2a17168ec895c781a9191c1e95ad7    274
dtype: int64

## Products Dataset — Initial Observations

- Each row represents one unique product in the product catalog.
- `product_id` uniquely identifies each product and is the primary key candidate.
- No exact duplicate rows were found.
- Missing values exist in product category and several product attribute columns.
- Missing product attributes should not be removed automatically because products with incomplete metadata may still have valid transactions.
- Every `product_id` referenced by Order Items exists in the Products dataset, indicating valid referential integrity from Order Items to Products.
- The Products to Order Items relationship is one-to-many because the same product can appear in multiple item rows across different orders.
- Product purchase frequency is highly uneven: some products are purchased many times while many products have relatively few purchases.
- Product category, weight, and dimensions may be useful later for product-mix, freight-cost, and profitability analysis.
- Missing categories will require a documented handling rule before category-level analysis.